### ISSUE TITLE
Standardization/grid issue

### ISSUE THREAD

**Issue by nlapier2** [stephenslab/mr.ash.alpha#7]
tl;dr -- MR.Ash seems to do much better when I regress GWAS Z-scores on other Z-scores versus betas on betas, and I am unsure if this is expected. Also, the default grid throws up warnings.

I have recently been using MR.Ash to perform regression in a Mendelian Randomization setting, in which I regress summary statistics for one phenotype on the summary statistics of other phenotypes. First, I simulate a genotype matrix G (actually just MVN in the code below), a phenotype matrix X, and a phenotype vector Y. Then I regress the phenotypes on the genotypes and retain the effect sizes (betas), i.e. summary statistics that one would get from a GWAS. Then I regress the snp effects on Y (y_betas) against the SNP effects on X (x_betas).

Here's some code to get the "summary statistics":

```
set.seed(1)
eff_size = 0.1
N=10000
M=100
K=30
theta_gx = matrix(rnorm(M*K), M, K)
theta_xy = c(rep(eff_size, 4), rep(0, K-4))

G = matrix(rnorm(N*M), N, M)
X = G %*% theta_gx + matrix(rnorm(N*K), N, K)
Y = X %*% theta_xy + rnorm(N)

regr_res = get_betas_stderrs(G, X, Y)
x_betas = regr_res$x_betas
x_stderrs = regr_res$x_stderrs
y_betas = regr_res$y_betas
y_stderrs = regr_res$y_stderrs
```

where get_betas_stderrs is a simple helper function to regress each X & Y on each "SNP" in G (sorry this code is a bit clunky):

```
get_betas_stderrs = function(G, X, Y) {
  num_snps = dim(G)[2]
  num_exposures = dim(X)[2]
  y_betas = integer(num_snps)
  y_stderrs = integer(num_snps)
  x_betas = matrix(0, num_exposures, num_snps)
  x_stderrs = matrix(0, num_exposures, num_snps)
  for (i in 1:num_snps) {
    Gi = G[, i]
    Gi = (Gi - mean(Gi)) / sd(Gi)  # standardize genotypes
    res_gy = summary(lm(Y ~ Gi))
    y_betas[i] = res_gy$coefficients[2,1]
    y_stderrs[i] = as.numeric(coef(res_gy)[, "Std. Error"][2])
    for (j in 1:num_exposures) {
      res_gx = summary(lm(X[, j] ~ Gi))
      x_betas[j,i] = res_gx$coefficients[2,1]
      x_stderrs[j,i] = as.numeric(coef(res_gx)[, "Std. Error"][2])
    }
  }
  x_betas = t(x_betas)
  x_stderrs = t(x_stderrs)
  return(list('x_betas' = x_betas, 'x_stderrs' = x_stderrs, 'y_betas' = y_betas, 'y_stderrs' = y_stderrs))
}
```

Then I run mr.ash to perform the regression of the summary statistics. When I run mr.ash(x_betas, y_betas) and compute the PIPs, there is little difference between the true effect variables (the first four -- see theta_xy above) and the other variables. However, if I compute the Zscores and run mr.ash on those -- mr.ash(z_x, z_y) -- it has much more signal. Code to replicate:

```
res.mr.ash = mr.ash(x_betas, y_betas)
full.post = get.full.posterior(res.mr.ash)
as.numeric(1 - full.post$phi[,1])

z_x = x_betas / x_stderrs
z_y = y_betas / y_stderrs
res.mr.ash = mr.ash(z_x, z_y)
full.post = get.full.posterior(res.mr.ash)
as.numeric(1 - full.post$phi[,1])
```

Is this expected? If running mr.ash on regression effect sizes, should they be standardized?

Also, note that the default grid throws a warning about a large amount of weight on the largest variance component. I did some tests with a much wider prior variance grid (code below). This gets rid of the warning, but the mixture weights look strange -- widely spread over many values, with weight (except for the 0 component) peaking at around sigma = 50 to 120. This seems strange to me. Any idea why this might be, or if it's related to the above issue?

```
grid = seq(from=0, to=200, by=1)
res.mr.ash = mr.ash(x_betas, y_betas, sa2=grid)
full.post = get.full.posterior(res.mr.ash)
as.numeric(1 - full.post$phi[,1])
as.numeric(res.mr.ash$pi)
```

**Comment by stephens999** [comment]
it isn't to do with standardizing because just dividing everything by 0.1 produces results that are different to not so dividing.
```
res.mr.ash = mr.ash(x_betas/0.1, y_betas/0.1)
full.post = get.full.posterior(res.mr.ash)
as.numeric(1 - full.post$phi[,1])
```
looks more like the z score results. 

Something looks wrong because mr.ash should be largely invariant to scaling.

**Comment by stephens999** [comment]
it looks to me that maybe the error is in get.full.posterior.... maybe in the computation of phi? 

```
get.full.posterior
function (fit) 
{
  r = fit$data$y - fit$data$X %*% fit$beta
  bw = as.vector((t(fit$data$X) %*% r) + fit$data$w * fit$beta)
  s2 = fit$sigma2/outer(fit$data$w, 1/fit$data$sa2, "+")
  m = bw * s2
  phi = -log(1 + outer(fit$data$w, fit$data$sa2))/2 + m * (bw/2/fit$sigma2)
  phi = c(fit$pi) * t(exp(phi - apply(phi, 1, max)))
  phi = t(phi)/colSums(phi)
  return(list(phi = phi, m = m, s2 = s2))
}
```

**Comment by pcarbo** [comment]
@stephens999 Indeed, see Issue #3.

**Comment by pcarbo** [comment]
@nlapier2 Your example wasn't fully reproducible; you forgot to set the seed. So I added that. 

When I ran the code, I found that the standard errors did not vary much, so the rescaling should have little impact on the final results:

```
> range(x_stderrs)
[1] 0.0864900 0.1193008
> range(y_stderrs)
[1] 0.02135971 0.02207812
```

**Comment by pcarbo** [comment]
@stephens999 @nathpgill I think the issue with this comparison is that the "spike" posterior probability `full.post$phi[,1]` is very sensitive to the choice of the prior variances; you will see that `data$sa2` is different for both runs of `mr.ash`. So I'm not sure comparing `phi[,1]` is all that useful.

**Comment by stephens999** [comment]
i think the issue was the previously-reported bug #3 get.full.posterior
I have fixed this. I think this will work now but have not checked. @nlapier2 can you try?

(The bug was in computation of the mean; computation of phi depends on the mean so this was also affected. The computation of phi is weird and obtuse, but i think it is likely correct once the mean is properly computed).

**Comment by pcarbo** [comment]
I confirm that there was a bug before, and matches the output from `varbvsmix` in this example:

```r
set.seed(1)
n <- 800
p <- 100
dat <- mr.ash::simulate_regression_data(n,p,s = 20)
X <- dat$X
y <- dat$y/3
sa   <- c(0,0.1,0.5,1,2,5)
fit1 <- mr.ash(X,y,sa2 = sa^2,sigma = 1,update.sigma2 = TRUE,
              intercept = TRUE,standardize = FALSE)
fit2 <- varbvs::varbvsmix(X,NULL,y,sa = sa^2,sigma = fit1$sigma,
                          w = fit1$pi,update.sigma = FALSE,
                          update.sa = FALSE,update.w = FALSE,
                          verbose = FALSE)
res  <- get.full.posterior(fit1)
print(range(res$phi - fit2$alpha))
# [1] -1.668273e-06  1.257439e-06
```

**Comment by nlapier2** [comment]
> i think the issue was the previously-reported bug #3 get.full.posterior I have fixed this. I think this will work now but have not checked. @nlapier2 can you try?
> 
> (The bug was in computation of the mean; computation of phi depends on the mean so this was also affected. The computation of phi is weird and obtuse, but i think it is likely correct once the mean is properly computed).

@stephens999 Yes, it works as expected now -- no difference after dividing by 0.1.
